File: omigamax/rules/board.py
"""Board state and the public Go playing API (todo 3).

Coordinates are ``(row, col)`` tuples, both 0-based, with ``(0, 0)`` the
top-left corner; a pass move is ``None``; colors are ``EMPTY`` (0),
``BLACK`` (1), ``WHITE`` (2).

Todo 4 (ko / two-pass terminal / scoring) and todo 5 (SGF) build on this
module -- the recorded history (``moves``, ``pass_count``,
``last_captured_point``) is the seam they slot into, so no breaking API
change is expected.
"""

from .captures import capture
from .ko import is_ko_prohibited
from .legality import IllegalMoveError, is_legal_move
from .liberties import (
    BLACK,
    EMPTY,
    WHITE,
    group,
    has_liberty,
    liberties,
    liberty_count,
)
from .scoring import (
    is_terminal,
    result_string,
    score,
    territory,
    winner,
)


class Board:
    """A Go board with parameterized size (default 19)."""

    def __init__(self, size=19):
        self.size = size
        self._state = [EMPTY] * (size * size)
        # -- history recorded for later todos (ko / terminal / SGF) --
        self.moves = []  # list of (move, color); move=None for a pass
        self.pass_count = 0  # consecutive passes since the last stone
        # Point (row, col) of a single-stone capture on the last move,
        # or None -- the data simple-ko detection (todo 4) needs.
        self.last_captured_point = None

    # -- accessors -----------------------------------------------------

    @property
    def state(self):
        """Copy of the flat board state (index ``r*size + c``)."""
        return list(self._state)

    def _idx(self, r, c):
        return r * self.size + c

    def is_on_board(self, r, c):
        """True if (r, c) lies within the board."""
        return 0 <= r < self.size and 0 <= c < self.size

    def get(self, r, c):
        """Color of the stone at (r, c) (``EMPTY`` if none)."""
        return self._state[self._idx(r, c)]

    def is_empty(self):
        """True if no stone has been placed yet."""
        return not any(self._state)

    # -- group / liberty queries (public API) --------------------------

    def group(self, r, c):
        """Set of coordinates of the connected group containing (r, c)."""
        return group(self._state, self.size, r, c)

    def liberties(self, r, c):
        """Set of empty points adjacent to the group containing (r, c)."""
        return liberties(self._state, self.size, r, c)

    def liberty_count(self, r, c):
        """Number of liberties of the group containing (r, c)."""
        return liberty_count(self._state, self.size, r, c)

    def has_liberty(self, r, c):
        """True if the group containing (r, c) has at least one liberty."""
        return has_liberty(self._state, self.size, r, c)

    # -- terminal / scoring (todo 4) ------------------------------------

    def is_terminal(self):
        """True if the game is over: two consecutive passes."""
        return is_terminal(self)

    def territory(self):
        """Map each empty point to its territory owner (see ``scoring``)."""
        return territory(self._state, self.size)

    def score(self, komi=7.5):
        """Tromp-Taylor area score ``(black, white)`` with komi on white."""
        return score(self._state, self.size, komi)

    def winner(self, komi=7.5):
        """``'B'``, ``'W'`` or ``None`` (jigo)."""
        return winner(self._state, self.size, komi)

    def result_string(self, komi=7.5):
        """SGF-style result, e.g. ``'B+3.5'`` or ``'W+2'``."""
        return result_string(self._state, self.size, komi)

    # -- play ----------------------------------------------------------

    def is_legal(self, move, color):
        """True if ``move`` is legal for ``color``.

        Pass (``None``) is always legal. Enforces bounds, occupancy, suicide
        prohibition (todo 3) and simple ko (todo 4): a move that would
        immediately retake the single stone captured on the opponent's last
        move is illegal. Superko is deliberately not enforced (plan's locked
        decision for todo 4).
        """
        if not is_legal_move(self._state, self.size, move, color):
            return False
        return not is_ko_prohibited(self.last_captured_point, move)

    def play(self, move, color, check_legal: bool = True):
        """Place a stone (or pass) and remove captured opponent stones.

        Returns the number of stones captured (0 for a pass or a
        non-capturing move). Raises :class:`IllegalMoveError` if ``move`` is
        not legal; the board is left unchanged in that case.

        ``check_legal=True`` (default) verifies legality first. Pass
        ``check_legal=False`` to skip the re-check when the caller already
        knows the move is legal (MCTS expansion builds children from
        ``node.legal_moves``, which were vetted by ``legal_actions`` -- the
        re-check doubles the per-child legality work on the hot path, P11
        self-play speedup). The caller guarantees legality in that case.
        """
        if move is None:
            self.moves.append((None, color))
            self.pass_count += 1
            self.last_captured_point = None
            return 0
        if check_legal and not self.is_legal(move, color):
            raise IllegalMoveError(
                f"illegal move {move} for color {color}"
            )
        r, c = move
        self.pass_count = 0
        self._state[self._idx(r, c)] = color
        removed, groups = capture(self._state, self.size, r, c, color)
        if removed == 1 and groups:
            # single-stone capture -> record its point for simple-ko (todo 4)
            self.last_captured_point = next(iter(groups[0]))
        else:
            self.last_captured_point = None
        self.moves.append((move, color))
        return removed

    def pass_move(self, color):
        """Play a pass move for ``color``."""
        return self.play(None, color)

File: cpp/board.cpp
// Implementation of the omigamax C++ rules core (see board.h).
//
// Every rule below replicates the Python reference exactly 鈥?the diff test
// tests/test_cpp_rules_diff.py asserts bit-exact equality against
// omigamax/rules/* on every move of 1000+ games plus boundary positions.

#include "board.h"

#include <cmath>

namespace omigamax {

namespace {

// Full group of the stone at (r, c) as flat indices; empty point -> empty.
// Mirrors liberties.py `group()` (DFS stack flood fill).
std::vector<int> group_of(const std::vector<std::int8_t>& state, int size,
                          int r, int c) {
    std::vector<int> result;
    const std::int8_t color = state[r * size + c];
    if (color == EMPTY) {
        return result;
    }
    std::vector<char> visited(static_cast<size_t>(size * size), 0);
    std::vector<int> stack;
    stack.reserve(static_cast<size_t>(size * size));
    stack.push_back(r * size + c);
    while (!stack.empty()) {
        const int cur = stack.back();
        stack.pop_back();
        if (visited[cur]) {
            continue;
        }
        visited[cur] = 1;
        result.push_back(cur);
        const int cr = cur / size;
        const int cc = cur % size;
        std::vector<std::pair<int, int>> nbrs;
        neighbors(cr, cc, size, nbrs);
        for (const auto& [nr, nc] : nbrs) {
            const int ni = nr * size + nc;
            if (state[ni] == color && !visited[ni]) {
                stack.push_back(ni);
            }
        }
    }
    return result;
}

// True if any opponent group adjacent to (r, c) has no liberties on `state`
// (which already has `color` placed at (r, c)). Same truthiness as
// captures.py `captured_groups()` being non-empty.
bool has_capturable_group(const std::vector<std::int8_t>& state, int size,
                          int r, int c, int color) {
    const std::int8_t opp = (color == BLACK) ? WHITE : BLACK;
    std::vector<char> seen(static_cast<size_t>(size * size), 0);
    std::vector<std::pair<int, int>> nbrs;
    neighbors(r, c, size, nbrs);
    for (const auto& [nr, nc] : nbrs) {
        const int ni = nr * size + nc;
        if (seen[ni]) {
            continue;
        }
        if (state[ni] != opp) {
            continue;
        }
        if (has_liberty(state, size, nr, nc)) {
            continue;
        }
        const std::vector<int> stones = group_of(state, size, nr, nc);
        for (const int s : stones) {
            seen[s] = 1;
        }
        return true;
    }
    return false;
}

// Replicates Python f"{x:g}" for the only values TT scoring can produce:
// non-negative integers and half-integers (0.0 / 0.5 / 1.0 / 1.5 / ...).
std::string fmt_margin(double d) {
    const double rounded = std::floor(d);
    if (d == rounded) {
        return std::to_string(static_cast<long long>(rounded));
    }
    return std::to_string(static_cast<long long>(rounded)) + ".5";
}

}  // namespace

void neighbors(int r, int c, int size,
               std::vector<std::pair<int, int>>& out) {
    out.clear();
    if (r > 0) {
        out.emplace_back(r - 1, c);
    }
    if (r < size - 1) {
        out.emplace_back(r + 1, c);
    }
    if (c > 0) {
        out.emplace_back(r, c - 1);
    }
    if (c < size - 1) {
        out.emplace_back(r, c + 1);
    }
}

bool has_liberty(const std::vector<std::int8_t>& state, int size, int r,
                 int c) {
    const std::int8_t color = state[r * size + c];
    if (color == EMPTY) {
        return false;
    }
    std::vector<char> visited(static_cast<size_t>(size * size), 0);
    std::vector<int> stack;
    stack.reserve(static_cast<size_t>(size * size));
    stack.push_back(r * size + c);
    while (!stack.empty()) {
        const int cur = stack.back();
        stack.pop_back();
        if (visited[cur]) {
            continue;
        }
        visited[cur] = 1;
        const int cr = cur / size;
        const int cc = cur % size;
        std::vector<std::pair<int, int>> nbrs;
        neighbors(cr, cc, size, nbrs);
        for (const auto& [nr, nc] : nbrs) {
            const int ni = nr * size + nc;
            if (state[ni] == EMPTY) {
                return true;
            }
            if (state[ni] == color && !visited[ni]) {
                stack.push_back(ni);
            }
        }
    }
    return false;
}

Board::Board(int size)
    : size_(size), state_(static_cast<size_t>(size * size), EMPTY) {
    if (size <= 0) {
        throw std::invalid_argument("board size must be positive");
    }
}

std::int8_t Board::get(int r, int c) const { return state_[idx(r, c)]; }

bool Board::is_on_board(int r, int c) const {
    return 0 <= r && r < size_ && 0 <= c && c < size_;
}

bool Board::is_empty() const {
    for (const std::int8_t v : state_) {
        if (v != EMPTY) {
            return false;
        }
    }
    return true;
}

std::vector<std::pair<int, int>> Board::group(int r, int c) const {
    std::vector<std::pair<int, int>> out;
    const std::vector<int> stones = group_of(state_, size_, r, c);
    out.reserve(stones.size());
    for (const int i : stones) {
        out.emplace_back(i / size_, i % size_);
    }
    return out;
}

std::vector<std::pair<int, int>> Board::liberties(int r, int c) const {
    std::vector<std::pair<int, int>> result;
    const std::vector<int> stones = group_of(state_, size_, r, c);
    std::vector<char> seen(static_cast<size_t>(size_ * size_), 0);
    std::vector<std::pair<int, int>> nbrs;
    for (const int i : stones) {
        const int gr = i / size_;
        const int gc = i % size_;
        neighbors(gr, gc, size_, nbrs);
        for (const auto& [nr, nc] : nbrs) {
            const int ni = nr * size_ + nc;
            if (state_[ni] == EMPTY && !seen[ni]) {
                seen[ni] = 1;
                result.emplace_back(nr, nc);
            }
        }
    }
    return result;
}

int Board::liberty_count(int r, int c) const {
    return static_cast<int>(liberties(r, c).size());
}

bool Board::has_liberty(int r, int c) const {
    return omigamax::has_liberty(state_, size_, r, c);
}

std::map<int, int> Board::territory() const {
    std::map<int, int> own;
    std::vector<char> visited(static_cast<size_t>(size_ * size_), 0);
    for (int start = 0; start < size_ * size_; ++start) {
        if (visited[start] || state_[start] != EMPTY) {
            continue;
        }
        std::vector<int> region;
        std::vector<std::int8_t> bordering;
        std::vector<int> stack;
        stack.push_back(start);
        while (!stack.empty()) {
            const int cur = stack.back();
            stack.pop_back();
            if (visited[cur]) {
                continue;
            }
            visited[cur] = 1;
            region.push_back(cur);
            const int r = cur / size_;
            const int c = cur % size_;
            std::vector<std::pair<int, int>> nbrs;
            neighbors(r, c, size_, nbrs);
            for (const auto& [nr, nc] : nbrs) {
                const int ni = nr * size_ + nc;
                if (state_[ni] == EMPTY) {
                    stack.push_back(ni);
                } else {
                    bordering.push_back(state_[ni]);
                }
            }
        }
        // Region is a set-collection of bordering colors; owner is the single
        // distinct color, or neutral (0) when bordered by both / none.
        const std::int8_t first = bordering.empty() ? EMPTY : bordering[0];
        int owner = 0;
        if (!bordering.empty()) {
            bool single = true;
            for (const std::int8_t b : bordering) {
                if (b != first) {
                    single = false;
                    break;
                }
            }
            if (single) {
                owner = first;
            }
        }
        for (const int i : region) {
            own[i] = owner;
        }
    }
    return own;
}

std::pair<double, double> Board::score(double komi) const {
    int black_stones = 0;
    int white_stones = 0;
    for (const std::int8_t v : state_) {
        if (v == BLACK) {
            ++black_stones;
        } else if (v == WHITE) {
            ++white_stones;
        }
    }
    int neutral = 0;
    int black_territory = 0;
    int white_territory = 0;
    for (const auto& [i, owner] : territory()) {
        (void)i;
        if (owner == BLACK) {
            ++black_territory;
        } else if (owner == WHITE) {
            ++white_territory;
        } else {
            ++neutral;
        }
    }
    const double half = static_cast<double>(neutral) / 2.0;
    // Same left-to-right evaluation as scoring.py:64-84; every term is
    // exactly representable (integers + 0.5 increments + komi), so the
    // result is bit-exact vs Python's float arithmetic.
    const double black_total =
        static_cast<double>(black_stones + black_territory) + half;
    const double white_total =
        static_cast<double>(white_stones + white_territory) + half + komi;
    return {black_total, white_total};
}

std::string Board::winner(double komi) const {
    const auto [black, white] = score(komi);
    if (black > white) {
        return "B";
    }
    if (white > black) {
        return "W";
    }
    return "";
}

std::string Board::result_string(double komi) const {
    const auto [black, white] = score(komi);
    if (black > white) {
        return "B+" + fmt_margin(black - white);
    }
    if (white > black) {
        return "W+" + fmt_margin(white - black);
    }
    return "Jigo";
}

bool Board::is_legal_move(int r, int c, int color) const {
    if (!is_on_board(r, c)) {
        return false;
    }
    const int i = idx(r, c);
    if (state_[i] != EMPTY) {
        return false;
    }
    std::vector<std::pair<int, int>> nbrs;
    neighbors(r, c, size_, nbrs);
    for (const auto& [nr, nc] : nbrs) {
        if (state_[nr * size_ + nc] == EMPTY) {
            return true;  // fast path: stone keeps at least one liberty
        }
    }
    // Every neighbor occupied: legal iff the move captures at least one
    // opponent stone or the placed stone joins a group that keeps a liberty.
    // Simulate the placement on a copy (Python restores in `finally`).
    std::vector<std::int8_t> tmp = state_;
    tmp[i] = static_cast<std::int8_t>(color);
    if (has_capturable_group(tmp, size_, r, c, color)) {
        return true;
    }
    return omigamax::has_liberty(tmp, size_, r, c);
}

bool Board::is_legal(int r, int c, int color) const {
    if (!is_legal_move(r, c, color)) {
        return false;
    }
    // Simple-ko: the move may not immediately retake the single stone
    // captured on the opponent's last move (ko.py:14-28).
    if (last_captured_ >= 0 && idx(r, c) == last_captured_) {
        return false;
    }
    return true;
}

std::pair<int, std::vector<std::vector<int>>> Board::capture(int r, int c,
                                                             int color) {
    const std::int8_t opp = (color == BLACK) ? WHITE : BLACK;
    std::vector<std::vector<int>> groups;
    std::vector<char> seen(static_cast<size_t>(size_ * size_), 0);
    std::vector<std::pair<int, int>> nbrs;
    neighbors(r, c, size_, nbrs);
    for (const auto& [nr, nc] : nbrs) {
        const int ni = nr * size_ + nc;
        if (seen[ni]) {
            continue;
        }
        if (state_[ni] != opp) {
            continue;
        }
        if (omigamax::has_liberty(state_, size_, nr, nc)) {
            continue;
        }
        std::vector<int> stones = group_of(state_, size_, nr, nc);
        for (const int s : stones) {
            seen[s] = 1;
        }
        groups.push_back(std::move(stones));
    }
    int total = 0;
    for (const auto& g : groups) {
        for (const int s : g) {
            if (state_[s] != EMPTY) {
                state_[s] = EMPTY;
                ++total;
            }
        }
    }
    return {total, std::move(groups)};
}

int Board::play(int r, int c, int color, bool check_legal) {
    if (check_legal && !is_legal(r, c, color)) {
        throw IllegalMoveError("illegal move (" + std::to_string(r) + ", " +
                               std::to_string(c) + ") for color " +
                               std::to_string(color));
    }
    pass_count_ = 0;
    state_[idx(r, c)] = static_cast<std::int8_t>(color);
    auto [removed, groups] = capture(r, c, color);
    if (removed == 1 && !groups.empty()) {
        // Single-stone capture -> record its point for simple-ko. The group
        // has exactly one stone, so the point is unambiguous.
        last_captured_ = groups[0][0];
    } else {
        last_captured_ = -1;
    }
    ++move_count_;
    return removed;
}

int Board::pass_move(int color) {
    (void)color;
    ++pass_count_;
    last_captured_ = -1;
    ++move_count_;
    return 0;
}

std::vector<int> Board::legal_actions(int color) const {
    std::vector<int> out;
    for (int r = 0; r < size_; ++r) {
        for (int c = 0; c < size_; ++c) {
            if (is_legal(r, c, color)) {
                out.push_back(idx(r, c));
            }
        }
    }
    out.push_back(size_ * size_);  // pass index (always legal)
    return out;
}

void Board::set_stone(int r, int c, int color) {
    state_[idx(r, c)] = static_cast<std::int8_t>(color);
    pass_count_ = 0;
    last_captured_ = -1;
    move_count_ = 0;
}

void Board::load_position(const std::vector<std::int8_t>& state,
                          int move_count, int pass_count,
                          int last_captured_index) {
    if (static_cast<int>(state.size()) != size_ * size_) {
        throw std::invalid_argument(
            "state length does not match board size");
    }
    state_ = state;
    move_count_ = move_count;
    pass_count_ = pass_count;
    last_captured_ = last_captured_index;
}

}  // namespace omigamax
